File: cpp/leetcode/017.cpp
class Solution {
  public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.empty()) {
            return res;
        }
        unordered_map<char, string> board{{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
        help(0, digits, board, "", res);
        return res;
    }

    void help(int n, string digits, unordered_map<char, string> &board, string ans, vector<string> &res) {
        if (n == digits.size()) {
            res.push_back(ans);
            return;
        }
        for (int i = 0; i < board[digits[n]].size(); i++) {
            // ans += board[digits[n]][i];
            help(n + 1, digits, board, ans + board[digits[n]][i], res);
            // ans.pop_back();
        }
    }
};

// Iterative
class Solution {
  public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.empty()) {
            return res;
        }
        unordered_map<char, string> board{{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
        res.push_back("");
        for (int i = 0; i < digits.size(); i++) {
            int n = res.size();
            for (int j = 0; j < n; j++) {
                string tmp = res.front();
                res.erase(res.begin());
                for (int k = 0; k < board[digits[i]].size(); k++) {
                    res.push_back(tmp + board[digits[i]][k]);
                }
            }
        }
        return res;
    }
};

File: python/leetcode/017.py
class Solution:
    def letterCombinations(self, digits):
        table = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
                 '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
        if len(digits) == 0:
            return []
        if len(digits) == 1:
            return list(table[digits[0]])
        pre = self.letterCombinations(digits[:-1])
        add = table[digits[-1]]
        return [s + c for s in pre for c in add]


class Solution(object):
    def letterCombinations(self, digits):
        res = []
        if not digits:
            return res
        table = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
                 '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
        self.dfs(table, digits, 0, "", res)
        return res

    def dfs(self, table, digits, n, ans, res):
        if n == len(digits):
            res.append(ans)
            return
        for ch in table[digits[n]]:
            self.dfs(table, digits, n + 1, ans + ch, res)
